### ISSUE TITLE
Support for adaptive average pooling?

### ISSUE THREAD

**Issue by paganpasta** [patrick-kidger/equinox#121]
Hi again,

I was trying out equinox for some computer vision experiments and found that at the moment there is no support for adaptive average pooling. A similar functionality exists in [Pytorch](https://pytorch.org/docs/stable/nn.html#pooling-layers). So I just wanted to check if it is something you intend to add later.

Thanks.

**Comment by patrick-kidger** [comment]
This is something I'd be happy to accept a PR on.

**Comment by paganpasta** [comment]
@patrick-kidger I'll definitely take a look at this. I'll first try to work on AdaptiveAvgPool1d and then make my way to 2d and 3d. 

Do you have any resources at hand regarding this? If not, don't worry, I found a Stack Overflow discussion ([link](https://stackoverflow.com/questions/53841509/how-does-adaptive-pooling-in-pytorch-work)) and will use it as a starting point.

**Comment by patrick-kidger** [comment]
I don't know of any resources I'm afraid. Good luck!

**Comment by paganpasta** [comment]
Hi @patrick-kidger 

I tried out a simple implementation for AdaptiveAvgPool1d to make sure that the overall working is correct. 

```
import jax.numpy as jnp
import equinox as eqx

#for comparing results
import torch  
import random


class AdaptiveAvgPool1d(eqx.Module):
    target_size: int

    def __init__(self, *,target_size):
        self.target_size = target_size
    
    def __call__(self, x):
        #Assertion based on https://pytorch.org/docs/stable/generated/torch.nn.AdaptiveAvgPool1d.html
        assert x.ndim == 2 or x.ndim==1, 'Only supports 2D:[row, cols] or 1D:[cols] input'  

        input_shape = x.shape
        start_points = (jnp.arange(self.target_size, dtype='float32')*(input_shape[-1]/self.target_size)).astype(dtype='int32')
        end_points = jnp.ceil((jnp.arange(self.target_size, dtype='float32')+1)*(input_shape[-1]/self.target_size)).astype(dtype='int32')
        pooled  = []
        for idx in range(self.target_size):
            if x.ndim == 1:
                pooled.append(jnp.mean(x[start_points[idx]:end_points[idx]], axis=-1, keepdims=False))
            else:
                pooled.append(jnp.mean(x[:,start_points[idx]:end_points[idx]], axis=-1, keepdims=False))
        
        x = jnp.asarray(pooled)
        if x.ndim == 2:
            x = jnp.swapaxes(jnp.asarray(pooled), -1, -2)
        return x


#### Test NxCxK ####
for i in range(100):

    n,c,k = random.randint(1, 10), random.randint(1, 10), random.randint(1, 10)
    pool_size = random.randint(1, 10)
    arr = torch.rand(n,c,k)
    
    torch_ans = torch.nn.AdaptiveAvgPool1d(pool_size)(arr)

    jnp_arr = jnp.asarray(arr.numpy().tolist())
    eq1d = AdaptiveAvgPool1d(target_size=pool_size)
    j1d = jax.vmap(eq1d)(jnp_arr)
    assert jnp.isclose(jnp.asarray(torch_ans.numpy().tolist()), j1d).all() == True, f'output mismatch for {n}x{c}x{k} with kernel {pool_size}' 


#### Test CxK ####
for i in range(100):

    c,k = random.randint(1, 10), random.randint(1, 10)
    pool_size = random.randint(1, 10)
    arr = torch.rand(c,k)
    
    torch_ans = torch.nn.AdaptiveAvgPool1d(pool_size)(arr)

    jnp_arr = jnp.asarray(arr.numpy().tolist())
    eq1d = AdaptiveAvgPool1d(target_size=pool_size)
    j1d = jax.vmap(eq1d)(jnp_arr)
    assert jnp.isclose(jnp.asarray(torch_ans.numpy().tolist()), j1d).all() == True, f'output mismatch for {c}x{k} with kernel {pool_size}'

```


At the moment I don't see how `eqx.nn.Pool` can be used as a base class for AdaptiveAvgPool1d. 
Also, the iteration bit of computing average can be optimized further?

**Comment by paganpasta** [comment]
```
class AdaptiveAvgPool1d(eqx.Module):
    target_size: int

    def __init__(self, *,target_size):
        self.target_size = target_size
    
    def __call__(self, x):
        assert x.ndim==1, 'Only supports 1D input' 
        channels = jnp.size(x)
        assert channels >= self.target_size, f'Final Pooled size {self.target_size} cannot be greater than input size {channels}'

        splits = jnp.array_split(x, self.target_size)
        num_head_arrays = channels%(self.target_size)
        if num_head_arrays:
            head_mean = jax.vmap(jnp.mean)(jnp.asarray(splits[:num_head_arrays]))
            tail_mean = jax.vmap(jnp.mean)(jnp.asarray(splits[num_head_arrays:]))
            mean = jnp.concatenate([head_mean, tail_mean])
        else:
            mean = jax.vmap(func)(jnp.asarray(splits))
        return mean
```
Agreed, there's no requirement to closely follow Pytorch implementation. I have added an extra check to make sure `final_size` < `num_channels`.

**Comment by paganpasta** [comment]
Merged in [v0.5.6](https://github.com/patrick-kidger/equinox/releases/tag/v0.5.6)
